src/eth_provider/database/ethereum.rs

```diff
@@ -356,4 +359,201 @@ mod tests {
+        test_get_transaction_count(&database, &mongo_fuzzer).await;
+    }
+
+    async fn test_get_header(database: &Database, mongo_fuzzer: &MongoFuzzer) {
+        let header_block_hash = mongo_fuzzer
+            .documents()
+            .get(&CollectionDB::Headers)
+            .unwrap()
+            .first()
+            .unwrap()
+            .extract_stored_header()
+            .unwrap()
+            .header
+            .clone();
+
+        let first_block_hash = header_block_hash.hash.unwrap();
+
+        // Test retrieving header by block hash
+        assert_eq!(database.header(first_block_hash.into()).await.unwrap().unwrap(), header_block_hash);
+
+        let first_block_number = header_block_hash.number.unwrap();
+
+        // Test retrieving header by block number
+        assert_eq!(database.header(first_block_hash.into()).await.unwrap().unwrap(), header_block_hash);
+
+        let mut rng = rand::thread_rng();
+        let false_block_hash = BlockHashOrNumber::from(B256::from(U256::from(rng.gen::<u64>())));
+
+        // Test retrieving non-existing header by block hash
+        assert_eq!(database.header(false_block_hash).await.unwrap(), None);
+
+        let false_block_number = BlockHashOrNumber::from(rng.gen::<u64>());
+
+        // Test retrieving non-existing header by block number
+        assert_eq!(database.header(false_block_number).await.unwrap(), None);
+    }
+
+    async fn test_get_blocks(database: &Database, mongo_fuzzer: &MongoFuzzer, u: &mut arbitrary::Unstructured<'_>) {
+        let header = mongo_fuzzer
+            .documents()
+            .get(&CollectionDB::Headers)
+            .unwrap()
+            .first()
+            .unwrap()
+            .extract_stored_header()
+            .unwrap()
+            .header
+            .clone();
+
+        let block_hash = header.hash.unwrap();
+
+        assert!(database
+            .block(block_hash.into(), true)
+            .await
+            .unwrap()
+            .unwrap()
+            .inner
+            .transactions
+            .as_transactions()
+            .is_some());
+
+        assert!(database
+            .block(block_hash.into(), false)
+            .await
+            .unwrap()
+            .unwrap()
+            .inner
+            .transactions
+            .as_hashes()
+            .is_some());
+
+        let first_block_number = header.number.unwrap();
+
+        assert!(database
+            .block(first_block_number.into(), true)
+            .await
+            .unwrap()
+            .unwrap()
+            .inner
+            .transactions
+            .as_transactions()
+            .is_some());
+
+        assert!(database
+            .block(first_block_number.into(), false)
+            .await
+            .unwrap()
+            .unwrap()
+            .inner
+            .transactions
+            .as_hashes()
+            .is_some());
```

Comment:
Here we must not just check that there is a `some` but to check that the content of the block returned is correct with an `assert_eq` compared to the block retrieved with the mongo fuzzer.

Reply:
I did that in a previous commit but was told off for using the same functions to build the block from the collection of transactions and headers from the database. 

Do you have a way of doing this in a consistent manner?

Reply:
In fact previously you were almost there with

```rust
let block: RichBlock = {
            let transactions = database.transactions(block_hash.into()).await.unwrap();
            let block_transactions = BlockTransactions::Hashes(transactions.iter().map(|tx| tx.hash).collect());

            let signed_transactions = transactions
                .into_iter()
                .map(|tx| TransactionSigned::try_from(tx).map_err(|_| EthereumDataFormatError::TransactionConversion))
                .collect::<Result<Vec<_>, _>>()
                .unwrap();

            let block = reth_primitives::Block {
                body: signed_transactions,
                header: reth_primitives::Header::try_from(header.clone())
                    .map_err(|_| EthereumDataFormatError::Primitive)
                    .unwrap(),
                withdrawals: Some(Default::default()),
                ..Default::default()
            };

            let size = block.length();

            Block {
                header: header.clone(),
                transactions: block_transactions,
                size: Some(U256::from(size)),
                withdrawals: Some(Default::default()),
                ..Default::default()
            }
            .into()
        };
```

Imo the only thing that is problematic here is that you use `database.transactions` and we don't want to use other implemented method here to isolate our tests. You could implement somewhere here

https://github.com/kkrt-labs/kakarot-rpc/blob/0e759fda9fc77e3b22f55cfbefc3b35cef6d4855/src/test_utils/katana/mod.rs#L286

a method `transactions_by_block_hash` to fetch the transactions by block hash in our mock data and not via the implementation. For the rest even if I'm not a big fan of tests which reimplement all the logic of the function, I must admit that we don't have too many other ways to rigorously transform the data without going through that especially for the size of the block